File: DailyChallenges/CPP/LeetCode_1402.cpp
class Solution {
public:
    int maxSatisfaction(vector<int>& satisfaction) {
        sort(satisfaction.begin(), satisfaction.end());
        int n = satisfaction.size();
        vector<int> dp(n+1, INT_MIN); dp[0] = 0;
        for (int i=1; i<=n; ++i){
            for (int j=i; j>0; --j)
                dp[j] = max(dp[j], dp[j-1] + j * satisfaction[i-1]);
        }
        return *max_element(dp.begin(), dp.end());
    }
};

File: DailyChallenges/Scala/LeetCode_1402.scala
object Solution {
    def maxSatisfaction(satisfaction: Array[Int]): Int = {
        val ss = satisfaction.sorted
        val dp = Array.fill[Int](ss.length+1)(Int.MinValue); dp(0) = 0
        for (i <- 1 to ss.length; j <- i until 0 by -1)
            dp(j) = dp(j) max (dp(j-1) + j * ss(i-1))
        dp.max
    }
}
